### PR TITLE
 setup test gems

### PR DESCRIPTION
**Pull request by Nicoale**
rspecrails, Factory bot and cuumber setup for testing

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+    validates :first_name, presence: true, length:{maximum: 30},format: { with: /\A[a-zA-Z]+\z/}
```
Comment (sigu) [inline app/models/user.rb:2]: Great, this is how to test using a regex 👍

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+    validates :last_name, presence: true, length:{maximum: 30}, format: { with: /\A[a-zA-Z]+\z/}
+    validates :phone_number, presence: true, uniqueness: true, length: {in: 10..13},format: { with: /\A[a-zA-Z]+\z/}
```
Comment (sigu) [inline app/models/user.rb:4]: Phone number needs to be a number not letters 😄 so the regex needs to be `/\d+/` 

`\d` means any digit
`+` means more than one

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    it { is_expected.to validate_format_of(:phone_number)} 
+    
+    it { is_expected.to validate_presence_of(:password)}
+    it { is_expected.to validate_length_of(:password)}
+    it { is_expected.to validate_presence_of(:pin)}
+    it { is_expected.to validate_length_of(:pin)}
+  end
+  it "raise database error when user attributes are null" do
+    valid_user_attrs = { first_name: "Jane", last_name: "Doe", id_number: 1234567, phone_number: "0700000000", pin: 1234, password: "PassWord@123."}
+    user_without_first_name_attr =   valid_user_attrs.slice(:last_name,:id_number,:phone_number, :pin, :password)
```
Comment (sigu) [inline spec/models/user_spec.rb:37]: What does `slice` do in this case?

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    user_without_first_name = User.new(user_without_first_name_attr)
+    expect { user_without_first_name.save!(validate: false)}.to raise_error(ActiveRecord::NotNullViolation)
```
Comment (sigu) [inline spec/models/user_spec.rb:39]: Do you understand what happened here? I am happy to explain in the afternoon call if you need to

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  end
+  describe "#creating user" do
+    it 'creates user with valid attributes' do
+      user = FactoryBot.create(:user) # we create user , insert him on the database
+      expect(user.first_name).to eq 'John' # we ensure his name is John since that is what is in the factory
+      expect(user.balance).to eq 0.0 # we assert that the user created has a balance of 0.0 remember default value of balance is 0.0
+      expect(user.pin).to eq 1234 # we assert that the pin for user created is 1234
+    end
+    it "do not allow existing phone number" do
+      FactoryBot.create(:user)
```
Comment (sigu) [inline spec/models/user_spec.rb:52]: What is the difference between ` FactoryBot.create(:user) ` and `FactoryBot.build(:user)` ?

File: spec/models/user_spec.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      expect(user.errors.messages[:phone_number]).to eq ["is not a number"] # # we test that error we get from phone number is "["is not a number"]"
+    end it "do not allow phone number that are less than 10" do
+      user = FactoryBot.build(:user, phone_number: "12345678")
+      user.save        
+      expect(user.errors.messages[:phone_number]).to eq ["is too short (minimum is 10 characters)"]
+    end
+    it "do not allow phone number that is greater than 13" do
+      user = FactoryBot.build(:user, phone_number: "12345678910112221")
+      user.save
+      expect(user.errors.messages[:phone_number]).to eq ["is too long (maximum is 13 characters)"]
```
Comment (sigu) [inline spec/models/user_spec.rb:71]: `user.errors.messages` is how we traverse errors in ruby on rails model objects.

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    user = FactoryBot.build(:user,id_number: "12345694")
+    user.save        
+    expect(user.errors.messages[:id_number]).to eq ["is invalid(the id number is already registered)"]
+  end
+  it "do not allow password characters that are less than 8" do
+    user = FactoryBot.build(:user, password: "123")
+    user.save        
+    expect(user.errors.messages[:password]).to eq ["is too short (minimum is 8 characters)"]
+  end
+  it "do not allow password characters that are greater than 8" do
```
Comment (sigu) [inline spec/models/user_spec.rb:98]: Make this to read more naturally "It does not allow password charaters that are greater than 8"

File: spec/models/user_spec.rb (hunk 6 above)
Comment (Nicoale) [reply spec/models/user_spec.rb:71]: oooh seen

File: app/models/user.rb (hunk 1 above)
Comment (Nicoale) [reply app/models/user.rb:2]: thank you

File: app/models/user.rb (hunk 2 above)
Comment (Nicoale) [reply app/models/user.rb:4]: noted will redo it

File: spec/models/user_spec.rb (hunk 3 above)
Comment (Nicoale) [reply spec/models/user_spec.rb:37]: slice is used to separate the attributes from the wanted ones. kind of used "to pick what you want "

File: spec/models/user_spec.rb (hunk 4 above)
Comment (Nicoale) [reply spec/models/user_spec.rb:39]: yes i  do

File: spec/models/user_spec.rb (hunk 5 above)
Comment (Nicoale) [reply spec/models/user_spec.rb:52]: with build you have to use .save

File: spec/models/user_spec.rb (hunk 7 above)
Comment (Nicoale) [reply spec/models/user_spec.rb]: okay noted and done.

**Pull request closed without merging**